BlueprintUICommonControls/Sources/ScrollView.swift

```diff
@@ -262,8 +304,120 @@ fileprivate final class ScrollerWrapperView: UIView {
+    
+    //
+    // MARK: UIView
+    //
+    
+    public override func didMoveToWindow() {
+        super.didMoveToWindow()
+        
+        if self.window != nil {
+            self.updateBottomContentInsetWithKeyboardFrame()
+        }
+    }
+    
+    public override func didMoveToSuperview() {
+        super.didMoveToSuperview()
+        
+        if self.superview != nil {
+            self.updateBottomContentInsetWithKeyboardFrame()
+        }
+    }
+}
 
 
+extension ScrollView
+{
+    // Calculates the correct content inset to apply for the given inputs.
+    
+    static func finalContentInset(
+        with providedInsets : UIEdgeInsets,
+        keyboardBottomInset : CGFloat,
+        refreshControlState : PullToRefreshBehavior,
+        refreshControlBounds : CGRect?
+    ) -> UIEdgeInsets
+    {
+        var finalContentInset = providedInsets
+        
+        // Include the keyboard's adjustment at the bottom of the scroll view.
+        
+        finalContentInset.bottom += keyboardBottomInset
+        
+        // The refresh control lives above the content and adjusts the
+        // content inset for itself when visible and refreshing.
+        // Do the same adjustment to our expected content inset.
+        
+        if case .refreshing = refreshControlState {
+            finalContentInset.top += refreshControlBounds?.size.height ?? 0.0
+        }
+        
+        return finalContentInset
     }
+}
 
+
+extension ScrollerWrapperView : KeyboardObserverDelegate {
+    
+    //
+    // MARK: Keyboard
+    //
+    
+    private func updateBottomContentInsetWithKeyboardFrame() {
+        
+        let contentInset = ScrollView.finalContentInset(
+            with: self.representedElement.contentInset,
+            keyboardBottomInset: self.bottomContentInsetAdjustmentForKeyboard,
+            refreshControlState: self.representedElement.pullToRefreshBehavior,
+            refreshControlBounds: self.refreshControl?.bounds
+        )
+        
+        /// Setting contentInset, even to the same value, can cause issues during scrolling (such as stopping scrolling).
+        /// Make sure we're only assigning the value if it changed.
+        
+        if self.scrollView.contentInset.bottom != contentInset.bottom {
+            self.scrollView.contentInset.bottom = contentInset.bottom
+        }
+    }
+    
+    fileprivate var bottomContentInsetAdjustmentForKeyboard : CGFloat {
+        
+        switch self.representedElement.keyboardAdjustmentMode {
+        case .none:
+            return 0.0
+            
+        case .adjustsWhenVisible:
+            guard let keyboardFrame = self.keyboardObserver.currentFrame(in: self) else {
+                return 0.0
+            }
+            
+            var bottomInset : CGFloat
+            
+            switch keyboardFrame {
+            case .notVisible: bottomInset = 0.0
+            case .visible(let frame): bottomInset = (self.bounds.size.height - frame.origin.y)
```

Comment:
Are you sure `bounds` is the correct height to use for this calculation, and not `frame`? If the content is taller than the frame of the scroll view, won't this calculation be incorrect? Maybe I misunderstand what's being calculated though 😅